Assets/MixedRealityToolkit/_Core/Sharing/Core/StateArray.cs

```diff
@@ -0,0 +1,395 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Pixie.Core
+{
+    // We use this class to get / set items in our sync list state in a standard way
+    public class StateArray<T> : IStateArray<T> where T : struct, IItemState, IItemStateComparer<T>
```

Comment:
Requiring T to be a struct seems limiting. Whats the thinking around requiring it to be a struct?

Reply:
Yes this is a deliberate limitation. The goal is to establish a central authority for state data. If you get a state directly from AppState then it's guaranteed to be an up-to-date copy. If you get it from anywhere else it may be stale / modified. And if you want to change a state you must do it through AppState for anyone else to be aware of those changes.

This all cuts down on the state 'spaghetti' you can get when objects are responsible for maintaining their own state data. It makes everything much easier to track and debug.

As a bonus it makes compression easier by encouraging the complete decoupling of state data from classes that use it.

Worth noting that these concerns don't come up in smaller apps. They can handle a bit of state spaghetti without an issue.

Reply:
Is there a good example in this review of a preexisting class that populates a state array?